mpa/seg/semisl/trainer.py

```diff
@@ -0,0 +1,165 @@
+        logger.info('Environment info:\n' + dash_line + env_info + '\n' + dash_line)
+
+        # Data
+        datasets = [build_dataset(cfg.data.train)]
+
+        # Dataset for HPO
+        hp_config = kwargs.get('hp_config', None)
+        if hp_config is not None:
+            import hpopt
+
+            if isinstance(datasets[0], list):
+                for idx, ds in enumerate(datasets[0]):
+                    datasets[0][idx] = hpopt.createHpoDataset(ds, hp_config)
+            else:
+                datasets[0] = hpopt.createHpoDataset(datasets[0], hp_config)
+
+        # Target classes
+        if 'task_adapt' in cfg:
+            target_classes = cfg.task_adapt.final
+        else:
+            target_classes = datasets[0].CLASSES
+
+        # Metadata
+        meta = dict()
+        meta['env_info'] = env_info
+        # meta['config'] = cfg.pretty_text
+        meta['seed'] = cfg.seed
+        meta['exp_name'] = cfg.work_dir
+        if cfg.checkpoint_config is not None:
+            cfg.checkpoint_config.meta = dict(
+                mmseg_version=__version__ + get_git_hash()[:7],
+                CLASSES=target_classes)
+
+        if distributed:
+            if cfg.dist_params.get('linear_scale_lr', False):
+                new_lr = len(cfg.gpu_ids) * cfg.optimizer.lr
+                logger.info(f'enabled linear scaling rule to the learning rate. \
+                    changed LR from {cfg.optimizer.lr} to {new_lr}')
+                cfg.optimizer.lr = new_lr
+
+        # Save config
+        # cfg.dump(os.path.join(cfg.work_dir, 'config.yaml'))
+        # logger.info(f'Config:\n{cfg.pretty_text}')
+
+        if distributed:
+            os.environ['MASTER_ADDR'] = cfg.dist_params.get('master_addr', 'localhost')
+            os.environ['MASTER_PORT'] = cfg.dist_params.get('master_port', '29500')
+            mp.spawn(SemiSegTrainer.train_worker, nprocs=len(cfg.gpu_ids),
+                     args=(target_classes, datasets, cfg, distributed, True, timestamp, meta))
+        else:
+            SemiSegTrainer.train_worker(
+                None,
+                target_classes,
+                datasets,
+                cfg,
+                distributed,
+                True,
+                timestamp,
+                meta
+            )
+
+        # Save outputs
+        output_ckpt_path = os.path.join(cfg.work_dir, 'latest.pth')
+        best_ckpt_path = glob.glob(os.path.join(cfg.work_dir, 'best_mDice_*.pth'))
+        if len(best_ckpt_path) > 0:
+            output_ckpt_path = best_ckpt_path[0]
+        best_ckpt_path = glob.glob(os.path.join(cfg.work_dir, 'best_mIoU_*.pth'))
+        if len(best_ckpt_path) > 0:
+            output_ckpt_path = best_ckpt_path[0]
+        return dict(final_ckpt=output_ckpt_path)
+
+    @staticmethod
+    def train_worker(gpu, target_classes, datasets, cfg, distributed=False, validate=False,
+                     timestamp=None, meta=None):
+        # logger = get_logger()
```

Comment:
Could you remove this line?

Reply:
Sure, I would remove it this time, other annotations will be removed in other refactoring PRs.